### PR TITLE
[2021-08-05] Libsodium post

### PR DESCRIPTION
**Pull request by bdemers**
This blog post has the following features:

- [x] A GitHub Repository with a polished README
- [x] A GitHub Repository under the github.com/oktadev account
- [x] A title that's approved by Dev Advocacy
- [x] A URL that's approved by Dev Advocacy
- [x] The content has been ran through Grammarly (https://www.grammarly.com/)
- [x] Rendered locally and confirmed that no Markdown typos exist

### CODE REVIEW

File: _source/_posts/2021-08-05-libsodium-encryption-go-python.md [hunk 1]
```diff
@@ -0,0 +1,340 @@
...
+- "Learn how to encrypt cross-language with Python and Go, using libsodium🧂"
+image: blog/libsodium-encryption-go-python/social.png
+type: awareness
+---
+
+The Networking and Cryptography library (NaCl pronounced "salt") is a software library that provides the core operations required to build cryptographic tools. Sodium is a fork of NaCl with an extended API; it's portable, and binaries are available to be used by various programming languages and operating systems. It comes in the form of a library called `libsodium`.
+
+Although there are several Python and Go cryptography libraries, it is primarily a matter of personal choice which one to use. One advantage of `libsodium` is that it is available for many languages.
+
+Today, we will use `libsodium` for Python to encrypt messages and decrypt them using `libsodium` in Go. Let's get started!
```
Comment (havi) [inline _source/_posts/2021-08-05-libsodium-encryption-go-python.md:20]: is there any specific reason why you switched prepositions ("for Python" vs "in Go")? I would tend to keep it consistent within the sentence (for, for) or (in, in), but maybe there's a nuance i'm missing?, or maybe it was grammarly's lame suggestion...? Optional tweak. Not a hard edit.

File: _source/_posts/2021-08-05-libsodium-encryption-go-python.md [hunk 2]
```diff
@@ -0,0 +1,340 @@
...
+
+Run the application to see the poem Jabberwocky restored:
+
+```bash
+go run decrypt.go
+```
+
+## What is Secret Key Encryption
+
+Secret key encryption uses a single key to both encrypt and decrypt a message. It is commonly used as a faster alternative to public-key encryption. The key itself is a fixed-length random number. As there is only a single key, both the sender and recipient need a copy of it. This requires a key exchange which is often performed out of band or by using public-key encryption.
```
Comment (havi) [inline _source/_posts/2021-08-05-libsodium-encryption-go-python.md:238]:
```suggestion
Secret-key encryption uses a single key to both encrypt and decrypt a message. It is commonly used as a faster alternative to public-key encryption. The key itself is a fixed-length random number. As there is only a single key, both the sender and recipient need a copy of it. This requires a key exchange which is often performed out of band or by using public-key encryption.
```

for consistency.

File: _source/_posts/2021-08-05-libsodium-encryption-go-python.md [hunk 3]
```diff
@@ -0,0 +1,340 @@
...
+
+## Encrypt a Message with `libsodium` and Secret Key
+
+We will again encrypt the poem Jabberwocky, but this time using secret key encryption. For convenience, we will again generate and store the key in a file:
```
Comment (havi) [inline _source/_posts/2021-08-05-libsodium-encryption-go-python.md:242]:
```suggestion
We will again encrypt the poem Jabberwocky, but this time using secret-key encryption. For convenience, we will again generate and store the key in a file:
```
consistency

File: _source/_posts/2021-08-05-libsodium-encryption-go-python.md [hunk 4]
```diff
@@ -0,0 +1,340 @@
...
+```bash
+python generate_keys.py alice
+python generate_keys.py bob
+```
+
+The files `key_alice_sk`, `key_alice_pk`, `key_bob_sk`, and `key_bob_pk` should have been created.
+
+## Encrypt a Message with `libsodium`
+
+The message to be encrypted is a text file containing the poem *Jabberwocky* by Lewis Carroll. The poem itself is somewhat cryptic, as the meaning of the words is open to interpretation! Rather fittingly, Jabberwocky was first published in the book *Through the Looking-Glass, and What Alice Found There*.
```
Comment (havi) [inline _source/_posts/2021-08-05-libsodium-encryption-go-python.md:131]:
```suggestion
The message to be encrypted is a text file containing the poem "Jabberwocky" by Lewis Carroll. The poem itself is somewhat cryptic, as the meaning of the words is open to interpretation! Rather fittingly, Jabberwocky was first published in the book *Through the Looking-Glass, and What Alice Found There*.
```
double-quotes

File: _source/_posts/2021-08-05-libsodium-encryption-go-python.md [hunk 5]
```diff
@@ -0,0 +1,340 @@
...
+
+Download the Jabberwocky poem to `jabberwocky.txt`:
```
Comment (havi) [inline _source/_posts/2021-08-05-libsodium-encryption-go-python.md:133]:
```suggestion
Download the "Jabberwocky" poem to `jabberwocky.txt`:
```
double-quotes

File: _source/_posts/2021-08-05-libsodium-encryption-go-python.md [hunk 6]
```diff
@@ -0,0 +1,340 @@
...
+
+func main() {
+	GetKeys()
+	Decrypt()
+}
+```
+
+Go is a strongly typed language, so we have to create 32-byte arrays to store the keys. The `GetKeys()` function reads Bob's private key and Alice's public key for the files. They are converted from hexadecimal strings into byte arrays and copied into the fixed-length arrays. The `Decrypt` function reads the encrypted message into a byte array. The nonce needs to be separated from the encrypted message and copied into a 24-byte array. The `box.Open()` function performs the decryption and then prints the decrypted message.
+
+Run the application to see the poem Jabberwocky restored:
```
Comment (havi) [inline _source/_posts/2021-08-05-libsodium-encryption-go-python.md:230]:
```suggestion
Run the application to see the poem "Jabberwocky" restored:
```

double-quotes

File: _source/_posts/2021-08-05-libsodium-encryption-go-python.md (hunk 3 above)
Comment (havi) [inline _source/_posts/2021-08-05-libsodium-encryption-go-python.md:242]:
```suggestion
We will again encrypt the poem "Jabberwocky," but this time using secret-key encryption. For convenience, we will again generate and store the key in a file:
```
double-quotes and hyphenate secret-key encryption

File: _source/_posts/2021-08-05-libsodium-encryption-go-python.md [hunk 7]
```diff
@@ -0,0 +1,340 @@
...
+
+func main() {
+	GetKey()
+	Decrypt()
+}
+```
+
+In this case, only the secret key is extracted from the file using the `GetKey()` function. The program reads the encrypted message from a file and extracts the nonce. Finally, the function `secretbox.Open()` decrypts the message using the nonce and the secret key.
+
+Run the program to see the Jabberwocky poem restored:
```
Comment (havi) [inline _source/_posts/2021-08-05-libsodium-encryption-go-python.md:326]:
```suggestion
Run the program to see the "Jabberwocky" poem restored:
```

**Review by havi — CHANGES_REQUESTED** [review]
with double quotes to all the Jabberwocky instances, and tweaked a couple other small bits.

File: _source/_posts/2021-08-05-libsodium-encryption-go-python.md (hunk 1 above)
Comment (bdemers) [inline _source/_posts/2021-08-05-libsodium-encryption-go-python.md:20]:
```suggestion
Today, we will use `libsodium` in Python to encrypt messages and decrypt them using `libsodium` in Go. Let's get started!
```

File: _source/_posts/2021-08-05-libsodium-encryption-go-python.md (hunk 3 above)
Comment (bdemers) [inline _source/_posts/2021-08-05-libsodium-encryption-go-python.md]:
```suggestion
We will again encrypt the poem Jabberwocky, but this time using secret-key encryption. For convenience, we will again generate and store the key in a file:
```

**Review by havi — APPROVED** [review]

**Pull request merged**